content/_code-samples/key-derivation/js/index.js

```diff
@@ -0,0 +1,219 @@
+            const decoded = hexToBytes(seedValue);
+            if(decoded.length === 16) {
+                return decoded
+            } else {
+                //Raise Error
+                throw new Error("incorrect decoded seed length")
+            }
+        }
+
+        //from rfc1751 Mnemonic
+        try {
+            const rippledMnemonic = rfc1751MnemonicToKey(seedValue)
+            if (rippledMnemonic.length === 16) {
+                return [...rippledMnemonic]
+            }
+        } catch (error) {
+            //try next seed mode
+        }
+
+        //from password seed
+        const encoded = encodeUTF8(seedValue)
+        return hashjs.sha512().update(encoded).digest().slice(0, 16);
+    }
+
+    _deriveED25519Keypair() {
+        const prefix = 'ED';
+        const rawPrivateKey = sha512half(this.bytes);
+        const privateKey = prefix + bytesToHex(rawPrivateKey);
+        const publicKey = prefix + bytesToHex(Ed25519.keyFromSecret(rawPrivateKey).pubBytes());
+        return { privateKey, publicKey };
+    }
+
+    _deriveSecp256K1Keypair(entropy, options) {
+        const prefix = '00';
+        const privateKey = prefix + this._deriveSecp256K1PrivateKey(entropy, options).toString(16, 64).toUpperCase()
+        const publicKey = bytesToHex(Secp256k1.keyFromPrivate(privateKey.slice(2))
+            .getPublic()
+            .encodeCompressed());
+        return { privateKey, publicKey };
+    }
+
+    _deriveSecp256K1PrivateKey(seed, opts = {}) {
+        const root = opts.validator;
+        const order = Secp256k1.curve.n;
+        // This private generator represents the `root` private key, and is what's
+        // used by validators for signing when a keypair is generated from a seed.
+        const privateGen = this._deriveScalar(seed);
+        if (root) {
+            // As returned by validation_create for a given seed
+            return privateGen;
+        }
+        const publicGen = Secp256k1.g.mul(privateGen);
+        // A seed can generate many keypairs as a function of the seed and a uint32.
+        // Almost everyone just uses the first account, `0`.
+        const accountIndex = opts.accountIndex || 0;
+
+        return this._deriveScalar(publicGen.encodeCompressed(), accountIndex)
+            .add(privateGen)
+            .mod(order);
+    }
+
+    _deriveScalar(bytes, discrim) {
+        const order = Secp256k1.curve.n;
+        for (let i = 0; i <= 0xffffffff; i++) {
+            // We hash the bytes to find a 256 bit number, looping until we are sure it
+            // is less than the order of the curve.
+            const hasher = new Sha512.default().add(bytes);
+            // If the optional discriminator index was passed in, update the hash.
+            if (discrim !== undefined) {
+                hasher.addU32(discrim);
+            }
+            hasher.addU32(i);
+            const key = hasher.first256BN();
+            /* istanbul ignore else */
```

Comment:
Why is this required?

Reply:
This function is taken directly from xrpl/ripple-keypairs, including the funny comments...